src/Glacier/MultipartUploader.php

```diff
@@ -0,0 +1,207 @@
+<?php
+namespace Aws\Glacier;
+
+use Aws\CommandInterface;
+use Aws\HashingStream;
+use Aws\Multipart\AbstractUploader;
+use Aws\Multipart\UploadState;
+use Aws\PhpHash;
+use Aws\ResultInterface;
+use GuzzleHttp\Psr7;
+use Psr\Http\Message\StreamableInterface as Stream;
+
+/**
+ * Encapsulates the execution of a multipart upload to Glacier.
+ */
+class MultipartUploader extends AbstractUploader
+{
+    protected function loadUploadWorkflowInfo()
+    {
+        $this->info = [
+            'command' => [
+                'initiate' => 'InitiateMultipartUpload',
+                'upload'   => 'UploadMultipartPart',
+                'complete' => 'CompleteMultipartUpload',
+                'abort'    => 'AbortMultipartUpload',
+            ],
+            'defaults' => [
+                'account_id' => '-',
+                'vault_name' => null,
+            ],
+            'id' => [
+                'account_id' => 'accountId',
+                'vault_name' => 'vaultName',
+                'upload_id'  => 'uploadId',
+            ],
+            'part' => [
+                'min_size' => 1048576,
+                'max_size' => 4294967296,
+                'max_num'  => 10000,
+                'param'    => 'range',
+            ],
+        ];
+    }
+
+    protected function loadStateFromService()
+    {
+        return $this->client->getPaginator('ListParts', $this->state->getId())
+            ->each(function (ResultInterface $result) {
+                // Get the part size from the first part in the first result.
+                if (!$this->state->getPartSize()) {
+                    $this->state->setPartSize($result['PartSizeInBytes']);
+                }
+                // Mark all the parts returned by ListParts as uploaded.
+                foreach ($result['Parts'] as $part) {
+                    list($rangeIndex, $rangeSize) = $this->parseRange($part['RangeInBytes']);
+                    $this->state->markPartAsUploaded($rangeIndex, [
+                        'size'     => $rangeSize,
+                        'checksum' => $part['SHA256TreeHash'],
+                    ]);
+                }
+            })
+            ->then(function () {
+                // Set the state's status, then return it to fulfill the promise.
+                $this->state->setStatus(UploadState::INITIATED);
+                return $this->state;
+            });
+    }
+
+    protected function determinePartSize()
+    {
+        // Make sure the part size is set.
+        $partSize = $this->config['part_size'] ?: $this->info['part']['min_size'];
+
+        // Calculate list of valid part sizes.
+        static $validSizes;
```

Comment:
Probably. Stupid closure bindings. If I make the closure static, it should fix that, right?